### PR TITLE
[SofaNG] Setup LinearSolver

### PR DESCRIPTION
**Pull request by fredroy**
and lots of cleaning.

#1527

Based on master

Deprecates SofaBaseLinearSolver, SofaGeneralLinearSolver, SofaPreconditioner, SofaDenseSolver, SofaSparseSolver.

Re-organise Linear Solvers into 3 categories:
 - Direct
 - Iterative
 - Preconditioner

Move Implementations of MatrixAccessors along its API in SofaCore (could be the occasion to definitely ditch CRSMultiMatrixAccessor), as they are not components + the utility class BlocMatrixWriter
And cleaning (namespaces, pragma, explicite instanciations, symbol export for windows, SReal instead of double, etc)

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by fredroy** [comment]
[ci-build][force-full-build][with-all-tests]

**Comment by alxbilger** [comment]
Would it be the right time to put metis in a separate module in order to link it in a plugin? If you agree with this approach of course.

**Comment by fredroy** [comment]
> Would it be the right time to put metis in a separate module in order to link it in a plugin? If you agree with this approach of course.

This is a really good idea, but it will take a little time so I think it would be better to do it in an other PR 🤠

File: Component/LinearSolver/Iterative/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    ${SOFACOMPONENTLINEARSOLVERITERATIVE_SOURCE_DIR}/init.h
+    ${SOFACOMPONENTLINEARSOLVERITERATIVE_SOURCE_DIR}/CGLinearSolver.h
+    ${SOFACOMPONENTLINEARSOLVERITERATIVE_SOURCE_DIR}/CGLinearSolver.inl
+    ${SOFACOMPONENTLINEARSOLVERITERATIVE_SOURCE_DIR}/GraphScatteredTypes.h
+    ${SOFACOMPONENTLINEARSOLVERITERATIVE_SOURCE_DIR}/MatrixLinearSolver.h
+    ${SOFACOMPONENTLINEARSOLVERITERATIVE_SOURCE_DIR}/MatrixLinearSolver.inl
+    ${SOFACOMPONENTLINEARSOLVERITERATIVE_SOURCE_DIR}/MinResLinearSolver.h
+    ${SOFACOMPONENTLINEARSOLVERITERATIVE_SOURCE_DIR}/MinResLinearSolver.inl
+    ${SOFACOMPONENTLINEARSOLVERITERATIVE_SOURCE_DIR}/ShewchukPCGLinearSolver.h
+    ${SOFACOMPONENTLINEARSOLVERITERATIVE_SOURCE_DIR}/ShewchukPCGLinearSolver.inl
```
Comment (alxbilger) [inline Component/LinearSolver/Iterative/CMakeLists.txt:17]: I see indentation inconsistency in this file on github

File: Component/LinearSolver/Direct/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+set(HEADER_FILES
+    ${SOFACOMPONENTLINEARSOLVERDIRECT_SOURCE_DIR}/config.h.in
+    ${SOFACOMPONENTLINEARSOLVERDIRECT_SOURCE_DIR}/init.h    
+    ${SOFACOMPONENTLINEARSOLVERDIRECT_SOURCE_DIR}/AsyncSparseLDLSolver.h
+    ${SOFACOMPONENTLINEARSOLVERDIRECT_SOURCE_DIR}/AsyncSparseLDLSolver.inl
+    ${SOFACOMPONENTLINEARSOLVERDIRECT_SOURCE_DIR}/BTDLinearSolver.h
+    ${SOFACOMPONENTLINEARSOLVERDIRECT_SOURCE_DIR}/BTDLinearSolver.inl
+    ${SOFACOMPONENTLINEARSOLVERDIRECT_SOURCE_DIR}/CholeskySolver.h
+    ${SOFACOMPONENTLINEARSOLVERDIRECT_SOURCE_DIR}/CholeskySolver.inl
+    ${SOFACOMPONENTLINEARSOLVERDIRECT_SOURCE_DIR}/FillReducingOrdering.h
```
Comment (alxbilger) [inline Component/LinearSolver/Direct/CMakeLists.txt]: This one is not a solver. I suggest to move it in SofaMatrix

File: Component/LinearSolver/Direct/src/sofa/component/linearsolver/direct/SVDLinearSolver.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+
+namespace sofa::component::linearsolver::direct
+{
+
+using namespace sofa::linearalgebra;
+
+int SVDLinearSolverClass = core::RegisterObject("Linear system solver using a SVD decomposition of a dense matrix")
+        .add< SVDLinearSolver< FullMatrix<SReal>, FullVector<SReal> > >()
+        .add< SVDLinearSolver< FullMatrix<float>, FullVector<float> > >()
```
Comment (alxbilger) [inline Component/LinearSolver/Direct/src/sofa/component/linearsolver/direct/SVDLinearSolver.cpp]: Is `float` necessary?

File: Component/LinearSolver/Direct/src/sofa/component/linearsolver/direct/SparseCholeskySolver.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+#include <sofa/core/ObjectFactory.h>
+
+namespace sofa::component::linearsolver::direct
+{
+
+using namespace sofa::linearalgebra;
+
+int SparseCholeskySolverClass = core::RegisterObject("Direct linear solver based on Sparse Cholesky factorization, implemented with the CSPARSE library")
+        .add< SparseCholeskySolver< CompressedRowSparseMatrix<double>,FullVector<double> > >(true)
+        .add< SparseCholeskySolver< CompressedRowSparseMatrix<float>,FullVector<float> > >()
```
Comment (alxbilger) [inline Component/LinearSolver/Direct/src/sofa/component/linearsolver/direct/SparseCholeskySolver.cpp]: Is `float` necessary?

File: Component/LinearSolver/Preconditioner/src/sofa/component/linearsolver/preconditioner/BlockJacobiPreconditioner.cpp [hunk 5]
```diff
@@ -19,24 +19,17 @@
...
-namespace component
-{
-
-namespace linearsolver
+namespace sofa::component::linearsolver::preconditioner
 {
 
 using namespace sofa::linearalgebra;
 
 int BlockJacobiPreconditionerClass = core::RegisterObject("Linear solver based on a NxN bloc diagonal matrix (i.e. block Jacobi preconditioner)")
```
Comment (alxbilger) [inline Component/LinearSolver/Preconditioner/src/sofa/component/linearsolver/preconditioner/BlockJacobiPreconditioner.cpp]:
```suggestion
int BlockJacobiPreconditionerClass = core::RegisterObject("Linear solver based on a NxN block diagonal matrix (i.e. block Jacobi preconditioner)")
```

File: Component/LinearSolver/Direct/src/sofa/component/linearsolver/direct/SVDLinearSolver.cpp (hunk 3 above)
Comment (fredroy) [reply Component/LinearSolver/Direct/src/sofa/component/linearsolver/direct/SVDLinearSolver.cpp]: It was here before but I can remove it, no prob

File: Component/LinearSolver/Direct/src/sofa/component/linearsolver/direct/SparseCholeskySolver.cpp (hunk 4 above)
Comment (fredroy) [reply Component/LinearSolver/Direct/src/sofa/component/linearsolver/direct/SparseCholeskySolver.cpp]: idem

**Comment by fredroy** [comment]
Setting to WIP, trying the installation process

Edit: good thing I tried, there was a problem in a cmake.in 🤥
Back to review

**Comment by fredroy** [comment]
[ci-build][force-full-build][with-all-tests]

**Review by epernod — APPROVED** [review]

**Comment by fredroy** [comment]
WIP -> rewriting history

EDIT: in the end, the only "non logical" rename by `git log ` is `modules/SofaPreconditioner/src/SofaPreconditioner/BlockJacobiPreconditioner.cpp => Component/Compat/src/SofaBaseLinearSolver/CGLinearSolver.h`

where BlockJacobiPreconditioner.cpp  is deleted and thinks it is now the (new file but temporary) compat file for CGLinearSolver.h (whose destiny is to be deleted anyway)

**Pull request merged**